README.md
# minetest
mods for minetest
Mycustom :
----------
    add some missing recipes like : craft darkage:chalk from it's powder, default:savana_dirt_with_dry_grass etc ;
                                  : cook cobble_condensed into 81 default:stone, grass into dry_grass, etc ;
    CAN link the nodes you want to 'shaper' mods like arcs, facade, circular_saw, cnc_machine, etc :
          For example, register darkage:chalk into pkarcs mod;
          register default:meselamp into technic_cnc;
          etc.
       The most powerfull : new shaped nodes keep all the abilities from their parents, like light source :
       a lot of shaping mods lost some fields from the original node but mycustom recover them after the mod done it's work.
    
    All is customizable, by the way of switches turnable in 'true' or 'false', individually or by branch;
    The settings made locally to a server by an admin remain valid while installing a new release : local settings are saved into worldpath
    A special log file 'mycustom_log.txt' is recreated each start in worldpath, with customizable level of details


releases.txt
MyCustom
a mod by gudule written firstly for the server 'Baobab Nation' to add some usefull recipes

Release 1: around 2021-05-25

	add cooking recipe for compressed cobble to give 9 stones (mod moreblocks);

Release 2: around 2021-05-28

	add optional_depends = darkage to the file mod.conf
	add crafting recipe to craft back darkage:chalk from 2 darkage:chalk_powder if the mod is installed

Release 3: 2021-06-01

	add this releases.txt file;
	in the file mod.conf, move "moreblocks" from 'depends' to 'optional_depends'
	add settings into init.lua to easily (de)activate recipes. BE AWARE de-activate a recipe will result in 'unknown' nodes and items ingame ! 
	add recipe for darkage:old_tuff_bricks;
	New functions :
	 can call pkarcs.register_node() with any node of any mod to make arcs from them :)
	 
Release 3.1 : 2021-06-05

	can call technic_cnc.register_all() for some nodes to be shaped by the cnc machine (mod technic_cnc)

Release 3.2 : 2021-06-07

	add cooking recipe for 1 extra:cobble_condensed to 81 default:stone;
	settings.Technic_cnc_must_register.main_switch set to false by default to avoid exceed limit of nodes in registered_nodes;

Release 3.3 : 2021-09-02

	add cooking recipe for compressed desert_cobble to give 9 desert_stones (mod moreblocks);

Release 3.4 : 2021-09-11

    add cooking recipes for some dirts that have no recipe yet, with default:<some ingredients> and optionally with <some_mod>:<some ingredients>

Release 4.0 : 2021-09-14
========================

    modify the structure of the mod :
    Now a release will not overwrite previous customs settings made localy by an admin of a server :
    local settings are saved in a file at
    <worldpath>/mycustom_local_settings.dat

release 4.0.1 : 2021-09-15

   bugfix and separate also recipes definitions from code to factory_recipes.lua file;
   copy (if not exist) mycustom_local_settings.lua and mycustom_local_recipes.lua from 
   mod path to world path to prevent release update to overwrite local settings made by the server's admin

release 4.0.2 : 2021-10-05

   modify the data structure in mycustom_local_settings.lua : tables of switches Pkarcs_must_register = {...} and
   Technic_cnc_must_register = {...} are both integrated in the structure 
   shapers = {
      pkarcs = {...},
      technic_cnc = {...},
   }
   a routine make the upgrade if needed;
   a routine checks the presence of mycustom_new_settings.lua to integrate new switches directly 
	in mycustom_local_settings.lua;
   added switches :
   in mycustom_default_settings.lua :
   -  max_nodes_allowed_by_minetest = 32767     -- the max number of nodes in minetest.registered_nodes;
	As this value would not change out of a change of the minetest's core engine, there is no raison to include
	this switch in "local_settings.lua" 
   in mycustom_local_settings.lua :
   -  node_places_to_left_free = (customizable) -- the min number of places we want mycustom to left free :
		after "mycustom" finish it work, be sure at least this amount of places will remain free for
 		a call to minetest.register_node().
   -  shapers[some_mod_name].subnodes_made = (adjustable) -- the number of nodes that will be created by this 
	particular mod when called for registering a node : for example, calling pkarcs to register a_node will 
	create 3 new nodes as <a_node_arc>, <a_node>_inner_arc and <a_node>_outer_arc.
	This number is adjustable because some servers has customized the amount of shapes enabled for some mods.
	This number is automatically adjusted after the first call of such a procedure to stay closest to the true;
	The call is done only if the precalculated number of new nodes is compatible with the free available places;
	
	To know those 3 numeric switches allow to avoid a server crash by precalculating the necessary amount of 
	free space in minetest.registered_nodes before calling a_mod.register(a_node).

Release 4.0.3 : 2021-10-12
	
	new shaper integrated in settings : circular_saw from mod moreblocks;

	when new settings are brought by a release (file <modpath>/mycustom_new_settings.lua exist), the player with 
	privs["server"]==true (the admin) receive a chat message 5sec after login; rename or remove the file will discard this message.
	why this message ? the new settings are integrated in <modpath> both in files .._default_settings.lua and .._local_settings.lua;
	that way, new settings (normaly unactivated) must be manually integrated in the <worldpath> local_settings.lua if customization
	is desired by the admin. the content of mycustom_new_settings.lua have explanations in comments and the right part of settings
	ready to copy-paste at the right place inside the file <worldpath>/mycustom_local_settings.lua.
	the file <modpath>/mycustom_local_settings.lua already integrate those new settings, so they will be ready to customize in case
	of missing file in <worldpath>.
	Why haven't i automatically integrated new settings in <worldpath>/mycustom_local_settings.lua ?
	- I no longer have any hair to pull out...
	So let the admin decide if he want to customize that new settings, at the price of a copy-paste :)

init.lua
--[[
Current release : 4.0.3

This mod was written by Gudule initialy for the use of the webserver 'baobab nation';
there is no licence attached except of course for all the mods reffered to ;
So feel free to use-modify-redistribute this mod as you want.

the main aim is to add some functionalities like 
	- usefull recipes forgotten by some installed mods;
	- link (register) some nodes to some mod to obtain shaped derivated nodes ;
see releases.txt for history

Some recipes can be activated / deactivated turning true / false the corresponding settings in the file
	<worldpath>\mycustom_local_recipes.lua  and 
some settings can be customized the same way in the file
	<worldpath>\mycustom_local_settings.lua

!! BE AWARE de-activate a recipe / a setting previously used may result in 'unknown' nodes and items ingame !!

!! BE AWARE if you add some reference to any mod : this referenced mod then MUST be writen in the "optional_depends"
	section of the file "mod.conf" !!
Without that, the corresponding action will abort.

--]]

myc = {}  -- spacename for easyest cleaning at end of work

myc.log_list = {  			-- available levels of log verbose
		[1] = "silent",		-- no log at all,
		[2] = "summary",  	-- only time elapsed at end of work
		[3] = "minimal", 	-- begin + end of major steps
		[4] = "details",	-- all actions a demanding administrator may want to be informed about
		[5] = "debug", 		-- all actions, step by step, for debug purposes
	}
myc.log_level = 4 			-- the level of explanations we want while executing code

-- get the start and stop time of mycustom to log the number of seconds used by this mod at load time :
myc.time_start = os.clock()
myc.time_elapsed = 0

local func_string = ""
local func_exec
local nb_ok = 0
local result = "" -- string to log result of registering a recipe

--[ load "default" settings : ]
dofile(minetest.get_modpath("mycustom").."/mycustom_default_settings.lua")

--[ load "default" recipes definitions : ]
dofile(minetest.get_modpath("mycustom").."/mycustom_default_recipes.lua")

--[ Register functions for mycustom : ]
dofile(minetest.get_modpath("mycustom").."/functions.lua")

log() --clear the file <world_path>/mycustom_log.txt (because of no parameter in the call log() )
if myc.log_level==1 then log(1,"silent mode.") end 
log(2,"mycustom start loading...") -- say hello :)

--[ Check for (and load) local settings made by the Admin of the server and update settings with those local settings : ]
check_local_settings()

--[ check for new settings to add to local settings : ]
check_new_settings()

--[[ check for local recipes definitions that would update the default recipes. 
	see file mycustom_default_recipes.lua for instructions to well manage the recipes ]]
check_local_recipes()

--[ Register recipes :]
last_output = "default:dirt_with_rainforest_litter"

for k, a_recipe_def in ipairs(recipes_defs) do

	result = "recipe #"..tostring(k)..": "
	if (a_recipe_def ~= nil) then
	if (a_recipe_def.ignore~=true) and
	   ((a_recipe_def.setting==nil) or (check_settings(a_recipe_def.setting)==true)) then -- if switches exist for this recipe, they are "on"

	   	if (check_depends(a_recipe_def.depends_of) == true) then -- depending mods are well installed
	   		
	   		if (a_recipe_def.recipe_is_function ~= nil) then 
	   			--[[ recipe is not an item but a function : typically the variable "last_output"
	   			so we will execute the function to rewrite the recipe : ]]

	   			--[[ avoid conflicts with field recipe, if somebody let recipe a table of classic recipe (yes i've done this mistake !) : ]]
	   			if (type(a_recipe_def.recipe)=="table") then
	   				recipes_defs[k].recipe_is_function = nil
	   			else
		   			func_string = "return "..a_recipe_def.def.recipe
	   				func_exec = loadstring(func_string)
					a_recipe_def.def.recipe = { {func_exec()}, }
				end
	   		end
	   		minetest.register_craft(a_recipe_def.def)
 			result = result.."registered (all conditions ok)"
	   		nb_ok = nb_ok+1

	   		-- for cycling crafts we may need to remember the last output of a recipe :
	   		if a_recipe_def.remember_last_output==true then last_output=a_recipe_def.def.output end

	   	else
	   		-- depends unsatisfied : some mods are not installed
	   		result = result.."aborted (at least 1 mod cited in depends is not installed)"
	   	end
	else
		-- switche(s) is/are "off" for this recipe or field ignore==true for this recipe
		if (a_recipe_def.ignore==true) then result = result.."aborted (field 'ignore'=true)" 
		else
			result = result.."aborted (in settings, at least 1 setting is false for that recipe)"
		end
	end 
	else
		result = result.."aborted (definition table is nil);"
	end -- recipe_def not nil
	log(4, result)
end
log(3, tostring(nb_ok).." recipes well registered ")


max_nodes_in_MT 	= settings.max_nodes_allowed_by_minetest or 32767 		-- should be considered as a constant 
nb_nodes_used		= count(minetest.registered_nodes) 						-- should be updated each time mycustom made a node registered_nodes
places_to_left_free = settings.node_places_to_left_free or 0 				-- How many free places in minetest.registered_nodes 
																			--   we DO WANT mycustom to left after it's end of work.
actual_nodes_left 	= max_nodes_in_MT - nb_nodes_used - places_to_left_free -- the free places for mycustom to register new nodes
mod_need_places 	= 1														-- nb of places needed by a shaper mod to register subnodes 


-- Now call some mods to register some nodes to shape them :
log(3, "Now call some mod functions to register nodes we want to be shaped by them :")
for mod_name, _ in pairs(settings.shapers) do
	register_to_mod(mod_name)
end
log(3, "all registered shaper mods called.")

-- Chat commands :
minetest.register_chatcommand("mycustom",{
	params = "?",
	description = "say the improvments mades by mycustom are readable in <world_path>/mycustom_log.txt",
	func = function(name)
		minetest.chat_send_player(name, "what the mod mycustom have done is readable in the file <world_path>/mycustom_log.txt")
	end,
})


-- After have done all the work to do, log the actual values of settings :
log(4, "\nafter mycustom has done it's work, the settings and recipes are :")
log(4, "\nmycustom settings :\n"..
 	   "-------------------\n"..dump(settings))
log(4, "\nmycustom recipes :\n"..
 	   "------------------\n")

for num, def in ipairs(recipes_defs) do
	if (def.recipe_is_function == true) then
		def.recipe_is_function = tostring(def.recipe_is_function).." (recipe was overwriten by the 'last_output' variable)"
	end
	if (def.ignore == true) then
		def.ignore = "true : that means this recipe had not been registered by mycustom into minetest engine"
	end
	log(4, "#"..tostring(num)..": "..dump(def))
end

-- say Good bye :)
::end_of_init::

myc.time_elapsed = os.clock()-myc.time_start
log(2, "Work done in "..string.format("%.3f", tostring(myc.time_elapsed)).." seconds.")
minetest.log("mycustom has loaded in "..string.format("%.3f", tostring(myc.time_elapsed)).." seconds. the work done is recorded in the file <world_path>/mycustom_log.txt")

-- clear the memory :
settings 			= nil
local_settings 		= nil
recipes_defs 		= nil
local_recipes_defs 	= nil
func_string 		= nil
func_exec   		= nil
nb_ok 				= nil
result 				= nil
max_nodes_in_MT 	= nil
nb_nodes_used		= nil
places_to_left_free = nil
actual_nodes_left 	= nil
mod_need_places 	= nil
myc 				= nil

mycustom_new_settings.lua
--[[	File mycustom_new_settings.lua : part of mycustom mod;
	release 4.0.3

	The goal of this file is to enable you to easily copy the lua code part at the right place in your <worldpath>/mycustom_local_settings.lua
	(if you have a previous release installed) to customize it.

	If you do not, those new settings will be taken in account as default settings.


	Below the hierarchical way to find the new settings in your file (if exist) <Worldpath>/mycustom_local_settings.lua :

local_settings = {
	shapers = {  ### And now the code to copy, below : ]]

		moreblocks = {				-- moreblocks provide circular_saw which is the targeted shaper. 
			main_switch = false,	-- if =false, no work at all will be done for this shaper, whatever the switches below.
			subnodes_made = 49,  	--[[ how many subnodes are made by this mod while registering a node. this value is used to test that actual
										registered nodes more that value will not exceed the limit of minetest.registered_nodes which is currently 32767;
										the shaper will be call for the 1st node with switch set to true, then the real amount of shaped mod will 
										be calculated and this value updated if needed.
										After that, if limit would be exceeded, the call will not be done to avoid server crash at load time. 
									 ]]
			mods = { 				-- Only if 'main_switch'==true, Check out the following mods to (not) link their nodes to the shaper: 
				extra = {
					activated = true,  						-- if false, no work will be done between the shaper and the nodes from this mod
					all_nodes = false,						-- if true, all nodes from this mod will be linked to the shaper
					nodes = {								-- nodes to take in account if 'all_nodes'==false :
						["extra:cobble_condensed"] = true,  -- if true, link this node to the shaper
					},
				},
				caverealms = {
					activated = true,  						-- if false, no work will be done between the shaper and the nodes from this mod
					all_nodes = false,						-- if true, all nodes from this mod will be linked to the shaper
					nodes = {								-- nodes to take in account if 'all_nodes'==false :
						["caverealms:glow_amethyst"] = true,  -- if true, link this node to the shaper
					},
				},				
			},
		},



mycustom_local_recipes.lua
--[[ file mycustom_local_recipes.lua 
	part of mycustom mod
	contains local definitions for recipes to add/modify the default recipes of mycustom mod.
	this file will not be erased by updates, so it will preserve your recipes.

	A single recipe is identified by it's UNIQUE numerical index.
	at load time, mycustom load this file and compare it against the mycustom's global variable recipes_defs :
	if the same index exist in both recipes_defs and recipes_defs_local, then 
		the recipes_defs_local will overwrite the recipes_def
	if you dont want a recipe anymore, you have three ways :
		1°) recommended way :
		--------------------
				create a field "ignore = true" in the recipe : this recipe will be ignore, like :
						recipes_defs_local = {
							[1] = {
					==>			ignore = true,
								description = "Cook moreblocks:cobble_compressed into 9 default:stone",
								depends_of = {"moreblocks",},
								setting = "Cook_compressed_cobble",
								def = {
									type = "cooking",
									output = "default:stone 9",
									recipe = "moreblocks:cobble_compressed",
								},
							},
							etc.
				this way you will remember easily what was this f**ing recipe n°1 AND you are sure this recipe will not be registered
		
		2°) you can turn to 'false' the corresponding setting in <worldpath>/mycustom_local_settings.dat;
			in the example above, you may do :  settings.Cook_compressed_cobble = false,
			but remember a setting may eventually apply to more than one recipe;
		
		3°) you can replace all the recipe's def by the value 'nil' like
				recipes_defs_local = {
					[1] = nil,
					[2] = ...
			but this barbarian method will lost what was the recipe n°1

		So, keep in mind if you work a different way than 1°) or 2°), you may run into bugs...

The differents fields findable in the recipe_defs_local :

	[index]		  : needed 		: Unique indentifier. DO NOT MODIFY, replace or erase. 
	"description" : needed 		: a short humain-readable description for this recipe; currently used in log messages.
	"depends_of"  : optional 	: if present, a table of strings representing mod(s) needed for that recipe to work.
								  if a mod isn't installed, discard the recipe.
	"setting"	  : optional	: if present, a (table of) string(s) representing the settings.<A_Switch_Name> to be true to activate the recipe
									for example, if setting = "Cook_compressed_cobble" then we must find settings.Cook_compressed_cobble==true to
									register this recipe.
									if setting is a table of strings, all the corresponding settings must be true
	"def"		  : needed		: a table containing the craft definition as minetest want it to be 
								 				will be used as minetest.register_craft(def)
	"ignore"	  : optional	: if present AND ==true, this recipe will not be registered whatever the other flags
	"remember_last_output"	: optional 	: if present, will save the "output" craft result in the global variable "last_output"
											this is used for cycling crafts for example for dirts_with_<something> crafted into <other_thing#1>,
											then <other_thing#2>, <#3>, ...<#n> and finally :
											<dirt_with_other_thing#n> cycled in craft <dirt_with_something>
	"recipe_is_function"	: optional : if present, the def.recipe is a string representing a lua code to execute as a function; 
											the result will overwrite the recipe def. 
											Typically the field def.recipe contains the string "last_output" then mycustom will replace it by 
											the value of last_output. 
											This mechanism let be possible the cycling crafts.


]]


local_recipes_defs = {
	[1] = {		-- numerical explicit index to preserve order because some recipes depends on previous recipes (e.g. cycling recipes for dirts)
		description = "Cook moreblocks:cobble_compressed into 9 default:stone",
		depends_of = {"moreblocks",},
		setting = "Cook_compressed_cobble",
		def = {
			type = "cooking",
			output = "default:stone 9",
			recipe = "moreblocks:cobble_compressed",
		},
	},
	[2] = {
		description = "Cook moreblocks:desert_cobble_compressed into 9 default:desert_stone",
		depends_of = {"moreblocks",},
		setting = "Cook_compressed_desert_cobble",
		def = {
			type = "cooking",
			output = "default:desert_stone 9",
			recipe = "moreblocks:desert_cobble_compressed",
		},
	},
	[3] = {
		description = "Cook extra:cobble_condensed into 81 default:stone",
		depends_of = {"extra",},
		setting = "Cook_condensed_cobble",
		def = {
			type = "cooking",
			output = "default:stone 81",
			recipe = "extra:cobble_condensed",
		},
	},
	[4] = {
		description = "Craft back darkage:chalk from 2 darkage:chalk_powder (horiz)",
		depends_of = {"darkage",},
		setting = "Craft_darkage_chalk",
		def = {
			output = "darkage:chalk",
			recipe = {
				{"darkage:chalk_powder",	"darkage:chalk_powder"},
			},
		},
	},
	[5] = {
		description = "Craft back darkage:chalk from 2 darkage:chalk_powder (vertic)",
		depends_of = {"darkage",},
		setting = "Craft_darkage_chalk",
		def = {
			output = "darkage:chalk",
			recipe = {
				{"darkage:chalk_powder"},
				{"darkage:chalk_powder"},
			},
		},
	},
	[6] = {
		description = "Craft darkage:old_tuff_bricks ",
		depends_of = {"darkage",},
		setting = "Craft_darkage_old_tuff_bricks",
		def = {
			output = "darkage:old_tuff_bricks 2",
			recipe = {
				{"darkage:mud",	"darkage:tuff_bricks"},
			},
		},
	},
	[7] = {
		description = "Craft darkage:old_tuff_bricks (alternate recipe)",
		depends_of = {"darkage",},
		setting = "Craft_darkage_old_tuff_bricks",
		def = {
			output = "darkage:old_tuff_bricks 2",
			recipe = {
				{"darkage:tuff_bricks", "darkage:mud"},
			},
		},
	},
	[8] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"default:pine_bush_needles"},
				{"default:dirt"},
			},
		},
	},
	[9] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"default:pine_needles"},
				{"default:dirt"},
			},
		},
	},
	[10] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"default:pine_bush_sapling"},
				{"default:dirt"},
			},
		},
	},
	[11] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"default:pine_bush_stem"},
				{"default:dirt"},
			},
		},
	},
	[12] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default", "trunks",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"trunks:pine_treeroot"},
				{"default:dirt"},
			},
		},
	},
	[13] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default", "moretrees",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"moretrees:cedar_cone"},
				{"default:dirt"},
			},
		},
	},
	[14] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default", "moretrees",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"moretrees:fir_cone"},
				{"default:dirt"},
			},
		},
	},
	[15] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default", "moretrees",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"moretrees:spruce_cone"},
				{"default:dirt"},
			},
		},
	},
	[16] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default", "moretrees",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"moretrees:acorn"},
				{"default:dirt"},
			},
		},
	},
	[17] = {
		description = "Cook default:dirt_with_grass into default:dry_dirt_with_dry_grass",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			type = "cooking",
			output = "default:dry_dirt_with_dry_grass",
			recipe = "default:dirt_with_grass",
			cooktime = 1,
		},
	},
	[18] = {
		description = "Craft default:dry_dirt_with_dry_grass",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dry_dirt_with_dry_grass",
			recipe = {
				{"default:dry_grass_1"},
				{"default:dry_dirt"},
			},
		},
	},
	[19] = {
		description = "Cook default:grass_1 into default:dry_grass_1",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			type = "cooking",
			output = "default:dry_grass_1",
			recipe = "default:grass_1",
			cooktime = 1,
		},
	},
	[20] = {
		description = "Cook default:dirt into default:dry_dirt",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			type = "cooking",
			output = "default:dry_dirt",
			recipe = "default:dirt",
			cooktime = 1,
		},
	},
	[21] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"default:jungleleaves"},
				{"default:dirt"},
			},
		},
	},
	[22] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"default:junglegrass"},
				{"default:dirt"},
			},
		},
	},
	[23] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"default:junglesapling"},
				{"default:dirt"},
			},
		},
	},
	[24] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default",},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"group:leaves"},
				{"default:dirt"},
			},
		},
		remember_last_output = true,
	},
	[25] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default","moretrees"},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"moretrees:jungletree_leaves_yellow"},
				{"default:dirt"},
			},
		},
	},
	[26] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default","moretrees"},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"moretrees:jungletree_leaves_red"},
				{"default:dirt"},
			},
		},
	},
	[27] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default","moretrees"},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"moretrees:jungletree_sapling_ongen"},
				{"default:dirt"},
			},
		},
	},
	[28] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default","vines"},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"vines:jungle_end"},
				{"default:dirt"},
			},
		},
	},
	[29] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default","vines"},
		setting = "Craft_dirts_default",
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"vines:jungle_middle"},
				{"default:dirt"},
			},
		},
	},
	[30] = {
		description = "Craft default:dirt_with_dry_grass",
		depends_of = {"default",},
		setting = {"Craft_dirts_default", "Craft_dirts_cycling",},
		def = {
			output = "default:dirt_with_dry_grass",
			recipe = {
				{"default:dirt_with_grass"},
			},
		},
	},
	[31] = {
		description = "Craft default:dry_dirt_with_dry_grass",
		depends_of = {"default",},
		setting = {"Craft_dirts_default", "Craft_dirts_cycling",},
		def = {
			output = "default:dry_dirt_with_dry_grass",
			recipe = {
				{"default:dirt_with_dry_grass"},
			},
		},
	},
	[32] = {
		description = "Craft default:dirt_with_coniferous_litter",
		depends_of = {"default",},
		setting = {"Craft_dirts_default", "Craft_dirts_cycling",},
		def = {
			output = "default:dirt_with_coniferous_litter",
			recipe = {
				{"default:dry_dirt_with_dry_grass"},
			},
		},
	},
	[33] = {
		description = "Craft default:dirt_with_rainforest_litter",
		depends_of = {"default",},
		setting = {"Craft_dirts_default", "Craft_dirts_cycling",},
		def = {
			output = "default:dirt_with_rainforest_litter",
			recipe = {
				{"default:dirt_with_coniferous_litter"},
			},
		},
	},
	[34] = {
		description = "Craft woodsoils:dirt_with_leaves_1",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod",},
		def = {
			output = "woodsoils:dirt_with_leaves_1",
			recipe = {
				{"group:leaves"},
				{"default:dirt_with_grass"},
			},
		},
	},
	[35] = {
		description = "Craft woodsoils:dirt_with_leaves_2",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod",},
		def = {
			output = "woodsoils:dirt_with_leaves_2",
			recipe = {
				{"woodsoils:dirt_with_leaves_1"},
			},
		},
	},
	[36] = {
		description = "Craft woodsoils:dirt_with_leaves_1 by cycling last_output",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod","Craft_dirts_cycling"},
		def = {
			output = "woodsoils:dirt_with_leaves_1",
			recipe = "last_output",
		},
		recipe_is_function = true,
	},
	[37] = {
		description = "Craft woodsoils:dirt_with_leaves_2",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod","Craft_dirts_cycling"},
		def = {
			output = "woodsoils:dirt_with_leaves_2",
			recipe = {
					{"woodsoils:dirt_with_leaves_1"},
				},
		},
	},
	[38] = {
		description = "Craft woodsoils:grass_with_leaves_1",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod",},
		def = {
			output = "woodsoils:grass_with_leaves_1",
			recipe = {
				{"group:leaves"},
				{"default:dirt_with_grass"},
			},
		},
	},
	[39] = {
		description = "Craft woodsoils:grass_with_leaves_2",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod",},
		def = {
			output = "woodsoils:grass_with_leaves_2",
			recipe = {
				{"woodsoils:grass_with_leaves_1"},
			},
		},
		remember_last_output = true,
	},
	[40] = {
		description = "Craft woodsoils:grass_with_leaves_1 by cycling",
		depends_of = {"default","woodsoils",},
		setting = {"Craft_dirts_mod","Craft_dirts_cycling"},
		def = {
			output = "woodsoils:grass_with_leaves_1",
			recipe = {
					{"woodsoils:dirt_with_leaves_2"},
				},
		},
	},
	[41] = {
		description = "Craft default:dirt_with_grass by cycling last_output",
		depends_of = {"default",},
		setting = {"Craft_dirts_default","Craft_dirts_cycling"},
		def = {
			output = "default:dirt_with_grass",
			recipe = "last_output",
			},
		recipe_is_function = true,
	},
}